design/authn-oidc.md

```diff
@@ -0,0 +1,121 @@
+
+## Overview
+
+Conjur supports various forms of authentication, including OIDC (OpenID Connect). OIDC is unique in that requires the user to
+navigate to a third party website (SSO provider) and authenticate there, often using MFA. The user then receives a token from
+the SSO provider and passes this to Conjur, which verifies that the token is signed correctly by the trust SSO provider.
+This document describes the design and implementation of the OIDC authentication support in the new Python-based Conjur CLI.
+
+### Motivation
+
+OIDC support was recently added to the Conjur UI (Enterprise only), initially to support SSO with Okta. We want to extend this
+support to the CLI so developers using the CLI can authenticate to Conjur in the CLI with OIDC as they will in the Conjur UI.
+
+
+## Solution
+
+### User Experience
+
+The `conjur init` command should support OIDC authentication as an authentication type. It should be possible to specify
+the authentication type `oidc` using the existing `-t` or `--authn-type` option.
+When using `oidc`, a `--service-id` option should be mandatory.
+
+*Open question*: Currently, if the `--service-id` option is specified, then `--authn-type` defaults to `ldap`. Should we make this required
+without a default value instead?
+
+| Option | Acceptable Values | Default | Description |
+| ------ | ----------------- | ------- | ----------- |
+| `--authn-type` / `-t` | `authn`, `ldap`, `oidc` | ??? | The authentication method to use when connecting to Conjur |
+| `--service-id` | `<service_id>` | N/A | The service_id of the OIDC service to use |
+
+The `conjur login` command should either:
+
+1. Open a browser to the OIDC URL configured for the SSO provider
+2. Print the URL (and/or a QR code) to be copied to a browser when running on a text-only machine
+
+The user must then authenticate to the SSO provider and receive a JWT which will be stored in the user's keychain or in the .netrc file.
+*TODO: How will the CLI get the JWT token from the browser?*
+Then when executing any CLI commands, the CLI will retrieve the stored JWT and pass it to Conjur's OIDC authentication endpoint and receive
+a short-lived accesss token, as with all other authentication strategies.
+
+#### Feature Details
+
+##### In the CLI repository
+
+The CLI needs to support the new CLI argument values in the `init` command, and it will need to pass them to the conjur-api-python library.
+Regarding the `login` step, this will be implemented very differently than the other authentication strategies (see details
+[below](#in-the-api-repository)).
+While this will mostly be contained in the API repository, there will need to be changes made to the login controller and logic
+in the CLI repository. For instance, when using authn-oidc, we won't be prompting the user for a username and password as this will
+be determined by the SSO provider.
```

Comment:
Plus any changes to .netrc (or keyring) which I assume is where the token(s) will be stored.

Reply:
I don't think this will require a change in the CLI project because the JWT will just replace the API token in the `CredentialsData` class.